PSF measurement, simulation, and deconvolution

talonchandler opened the pull request:
- paired with [this confluence page](https://czbiohub.atlassian.net/wiki/spaces/dashboard/pages/3411214430/2024-02-05+modeling+mantis+fluorescence+PSFs) and [this confluence page](https://czbiohub.atlassian.net/wiki/spaces/dashboard/pages/3567288346/2024-05-05+PSF+fitting+for+mantis+deconvolution) 
- 2024-05-07 - now includes draft CLI calls `psf-from-beads` which estimates an averaged PSF from a bead volume, and `deconvolve` which uses the estimated PSF to deconvolve

TODO:
 - [ ] update pre-commit

talonchandler commented on mantis/cli/deconvolve.py:
```diff
@@ -0,0 +1,123 @@
+            "Consider resampling the PSF."
+        )
+
+    for t in range(1):#T):
+        for c in range(C):
```
Kludge here...needs multiprocessing and `process_single_position`

talonchandler commented on mantis/cli/characterize.py:
```diff
@@ -0,0 +1,86 @@
+    output_dirpath: str,
+):
+    """
+    Characterize the point spread function (PSF) from bead images in an html report
+
+    >> mantis characterize -i ./beads.zarr/*/*/* -c ./characterize_params.yml -o ./
+    """
```
Only characterizes first position. Could use better docs.

ieivanov commented:
@talonchandler would you like to keep `scripts/fit_psf_to_beads.py`, `scripts/simple_psf.py`, and `scripts/simulate_psf.py`? They may need to have their deskew updated. These scripts call on `deskew/_deskew_matrix` which I think now is deprecated. Or is `deskew/_get_transform_matrix` the deprecated one? `deskew_data` uses
```
    matrix = _get_transform_matrix(
        raw_data.shape, ls_angle_deg, px_to_scan_ratio, keep_overhang
    )
```
but I think that may be a bug? I'd like to remove one of `_deskew_matrix` or `_get_transform_matrix`.

ieivanov commented:
P. S. If that's a bug in the deskew, let's fix it in a separate PR that we can merge right away

ieivanov commented:
> P. S. If that's a bug in the deskew, let's fix it in a separate PR that we can merge right away

This is a bug in this branch only, I'll fix it

talonchandler commented:
I removed `scripts/fit_psf_to_beads.py`, `scripts/simple_psf.py`, and `scripts/simulate_psf.py` because I don't think we'll need them. I can resurrect them if that changes. 

Those are the only places that `_deskew_matrix` is used, so I've deleted that function, too. 

With those deletions we can focus our attention on the `_get_transform_matrix`, which was regressing from the `main` branch (the GPU deskew branch changed the affine transformation matrix). I added a test to catch this problem, then fixed the problem. 

I spot-tested a deskew on real data, so I now think the deskew in this branch is ready (a minor refactor from `main`)

@ieivanov thanks for flagging this. I just took a look, and I think these `deskew` functions were the only places where accidental regressions happened.

ieivanov commented:
I'll test the refactored `measure_psf` today and then we can merge

ieivanov reviewed (approved):
I've tested `measure_psf` on mantis, I think this PR is ready to merge

talonchandler commented on mantis/cli/characterize_psf.py:
```diff
@@ -0,0 +1,113 @@
+    """
+    Characterize the point spread function (PSF) from bead images and output an html report
+
+    >> mantis characterize -i ./beads.zarr/*/*/* -c ./characterize_params.yml -o ./
```
```suggestion
    >> mantis characterize-psf -i ./beads.zarr/*/*/* -c ./characterize_params.yml -o ./
```

talonchandler commented on mantis/cli/estimate_psf.py:
```diff
@@ -0,0 +1,124 @@
+    """
+    Estimate the point spread function (PSF) from bead images
+
+    >> mantis psf_from_beads -i ./beads.zarr/*/*/* -c ./psf_params.yml -o ./psf.zarr
```
```suggestion
    >> mantis estimate-psf -i ./beads.zarr/*/*/* -c ./psf_params.yml -o ./psf.zarr
```

talonchandler commented:
LGTM!
